src/core/api/webhook_policy.go

```diff
@@ -0,0 +1,494 @@
+		return
+	}
+
+	oriPolicy, err := webhook.PolicyCtl.Get(id)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to get the webhook policy %d: %v", id, err))
+		return
+	}
+	if oriPolicy == nil {
+		w.SendNotFoundError(fmt.Errorf("webhook policy %d not found", id))
+		return
+	}
+
+	policy := &apiModels.WebhookPolicy{}
+	isValid, err := w.DecodeJSONReqAndValidate(policy)
+	if !isValid {
+		w.SendBadRequestError(err)
+		return
+	}
+
+	if !w.validateTargets(policy) {
+		return
+	}
+
+	if !w.validateHookTypes(policy) {
+		return
+	}
+
+	if w.project.ProjectID != oriPolicy.ProjectID {
+		w.SendBadRequestError(fmt.Errorf("webhook policy %d with projectID %d not belong to project %d in URL", id, oriPolicy.ProjectID, w.project.ProjectID))
+		return
+	}
+
+	ply, err := convertFromAPIModel(policy)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to convert webhook policy from api model: %v", err))
+		return
+	}
+	ply.ID = id
+	ply.ProjectID = w.project.ProjectID
+
+	if err = webhook.PolicyCtl.Update(ply); err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to update the webhook policy: %v", err))
+		return
+	}
+}
+
+// List ...
+func (w *WebhookPolicyAPI) List() {
+	projectID := w.project.ProjectID
+	if !w.validateRBAC(rbac.ActionList, projectID) {
+		return
+	}
+
+	_, res, err := webhook.PolicyCtl.List(projectID)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to list webhook policies by projectID %d: %v", projectID, err))
+		return
+	}
+
+	policies := []*apiModels.WebhookPolicy{}
+	if res != nil {
+		for _, policy := range res {
+			ply, err := convertToAPIModel(policy)
+			if err != nil {
+				w.SendInternalServerError(fmt.Errorf("failed to convert webhook policy to api model: %v", err))
+				return
+			}
+			policies = append(policies, ply)
+		}
+	}
+
+	w.WriteJSONData(policies)
+}
+
+// ListGroupByHookType lists webhook policy trigger info grouped by hook type for UI,
+// displays hook type, status(enabled/disabled), create time, last trigger time
+func (w *WebhookPolicyAPI) ListGroupByHookType() {
+	projectID := w.project.ProjectID
+	if !w.validateRBAC(rbac.ActionList, projectID) {
+		return
+	}
+
+	_, res, err := webhook.PolicyCtl.List(projectID)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to list webhook policies by projectID %d: %v", projectID, err))
+		return
+	}
+
+	policies, err := constructPolicyForUI(res)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to list the webhook policy trigger information: %v", err))
+		return
+	}
+	w.WriteJSONData(policies)
+}
+
+// Delete ...
+func (w *WebhookPolicyAPI) Delete() {
+	projectID := w.project.ProjectID
+	if !w.validateRBAC(rbac.ActionDelete, projectID) {
+		return
+	}
+
+	id, err := w.GetIDFromURL()
+	if id < 0 || err != nil {
+		w.SendBadRequestError(errors.New("invalid webhook policy ID"))
+		return
+	}
+
+	policy, err := webhook.PolicyCtl.Get(id)
```

Comment:
The checking should be moved to controller?

Reply:
> The checking should be moved to controller?

 I think checking should be done in api handler layer;
controller have been deleted